### PR TITLE
Implement built-ins

### PR DESCRIPTION
**Pull request by ianyong**
Necessary for `malloc`, `free`, etc.

### CODE REVIEW

File: src/interpreter/__tests__/builtins.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { parse } from '../../parser/parser';
+import { compileProgram } from '../compiler';
+import { interpret } from '../virtualMachine';
+
+describe('built-ins', () => {
+  test('handles built-in functions', () => {
+    const code = `
+        int main() {
+            return sqrt(25);
```
Comment (zognin) [inline src/interpreter/__tests__/builtins.ts:9]: What happens if I do `sqrt(7)`, aka if the result is non int? Does it matter? I think it doesn't really matter for our proof of concept but just wondering if you have any thoughts?

File: src/interpreter/compiler.ts [hunk 2]
```diff
@@ -419,10 +432,16 @@ const compilers: CompilerMapping = {
...
     jumpInstr.instrAddress = instructions.length;
 
-    const assignInstr = constructAssignInstr(
-      getSymbolTableEntry(getNameFromDeclaratorPattern(node.id), symbolTable),
-      1
+    const symbolTableEntry = getSymbolTableEntry(
+      getNameFromDeclaratorPattern(node.id),
+      symbolTable
     );
+    if (isBuiltinFunctionSymbolTableEntry(symbolTableEntry)) {
```
Comment (zognin) [inline src/interpreter/compiler.ts]: If we're checking whether the name belongs to a user declared function, would it be more comprehensive to check the symbol table entry type is `Function`?

File: src/interpreter/compiler.ts [hunk 3]
```diff
@@ -697,6 +721,11 @@ const compilers: CompilerMapping = {
         ) {
           throw new UnsupportedInitializationError();
         }
+        if (isBuiltinFunctionSymbolTableEntry(entry)) {
```
Comment (zognin) [inline src/interpreter/compiler.ts]: Similarly for this, why don't we check that the entry type is `Variable` and everything else should throw error?

**Review by zognin — APPROVED** [review]
LGTM, left some questions

File: src/interpreter/__tests__/builtins.ts (hunk 1 above)
Comment (ianyong) [reply src/interpreter/__tests__/builtins.ts:9]: You get a float. This is true of everything in our current implementation by the way. The following is a valid program:
```c
int main() {
    return 7.5;
}
```

I'm guessing we handle when type checking?

File: src/interpreter/compiler.ts (hunk 2 above)
Comment (ianyong) [reply src/interpreter/compiler.ts]: Oh yea, that's a good idea.

File: src/interpreter/compiler.ts (hunk 3 above)
Comment (ianyong) [reply src/interpreter/compiler.ts]: Apparently, the entry type can also be `Array` here?

File: src/interpreter/symbolTable.ts [hunk 4]
```diff
@@ -113,6 +116,11 @@ export const addFunctionSymbolTableEntries = (
     getNameFromDeclaratorPattern(node.id),
     symbolTable
   );
+  if (!isFunctionSymbolTableEntry(functionEntry)) {
```
Comment (zognin) [inline src/interpreter/symbolTable.ts]: I thought you only want user defined functions for this, and not builtins?

Comment (ianyong) [reply src/interpreter/symbolTable.ts]: This is the symbol table entry for user-defined functions. Should I rename `FunctionSymbolTableEntry` to `UserDeclaredFunctionSymbolTableEntry`?

Comment (zognin) [reply src/interpreter/symbolTable.ts]: oh okay anything

**Pull request merged**